### PR TITLE
feat: always initialize mediator

### PR DESCRIPTION
**Pull request by niallshaw-absa**
Add mediationRole property to AgentConfig to ensure MediatorModule is initialized correctly. This prevents concurrency issues when multiple attempts to create a mediatorRoutingRecord fail.

Signed-off-by: Niall Shaw <niall.shaw@absa.africa>

### CODE REVIEW

**Comment by niallshaw-absa** [comment]
Aims to fix #899

**Comment by TimoGlastra** [comment]
Technically an agent can be both a mediator and a recipient, so I'm not to keen on adding the role property to the config. I see two options, but there's probably other good options:
- Always initialize it. It adds some overhead, but it's not that much
- Add an option e.g. `initializeMediatorKeysOnStartup` that will init the mediator keys.

In addition, I'm not sure if it should be impossible to use mediation without the mediation role . E.g. with the ledger we have connectToLedgersOnStartup. If you have it enabled it will connect on startup, but otherwise when first requested. Is that something we could also do here?

Another issue is that this won't work for high availability deployments (multiple instances connecting to the same database). There's more places where this will lead to issues, so we can probably ignore this for now.

File: samples/mediator.ts [hunk 1]
```diff
@@ -37,7 +37,7 @@ const port = process.env.AGENT_PORT ? Number(process.env.AGENT_PORT) : 3001
 const app = express()
 const socketServer = new Server({ noServer: true })
 
-const endpoints = process.env.AGENT_ENDPOINTS?.split(',') ?? [`http://localhost:${port}`, `ws://localhost:${port}`]
+const endpoints = process.env.AGENT_ENDPOINTS?.split(',') ?? [`https://ebff-31-30-172-153.ngrok.io`]
```
Comment (jakubkoci) [inline samples/mediator.ts:40]: I'm not sure if we want to commit this change.

File: packages/core/src/agent/Agent.ts [hunk 2]
```diff
@@ -233,8 +235,8 @@ export class Agent {
       const mediationConnection = await this.getMediationConnection(mediatorConnectionsInvite)
       await this.mediationRecipient.provision(mediationConnection)
     }
-
-    await this.mediationRecipient.initialize()
+    if (mediationRole === MediationRole.Mediator) await this.mediator.initialize()
+    if (mediationRole === MediationRole.Recipient) await this.mediationRecipient.initialize()
```
Comment (jakubkoci) [inline packages/core/src/agent/Agent.ts:239]: Should not we also check if `mediationRole === MediationRole.Recipient` before we try to use `mediatorConnectionsInvite`?

**Comment by jakubkoci** [comment]
> Technically an agent can be both a mediator and a recipient

Then it might be an array of roles. I usually try to avoid these flag config attributes, but if the array is a problem I'm also ok with something like `initializeMediatorRoutingOnStartup`. The word `Keys` seem to be too low-level detail. 

> Another issue is that this won't work for high availability deployments (multiple instances connecting to the same database). There's more places where this will lead to issues, so we can probably ignore this for now.

Yes, without this change, it doesn't work even with a few edge agents trying to connect at the same time, so I would rather do this now and fix "high availability deployments" later.

**Comment by genaris** [comment]
Is there any 

> Technically an agent can be both a mediator and a recipient, so I'm not to keen on adding the role property to the config. I see two options, but there's probably other good options:
> 
> * Always initialize it. It adds some overhead, but it's not that much
> * Add an option e.g. `initializeMediatorKeysOnStartup` that will init the mediator keys.
> 
> In addition, I'm not sure if it should be impossible to use mediation without the mediation role . E.g. with the ledger we have connectToLedgersOnStartup. If you have it enabled it will connect on startup, but otherwise when first requested. Is that something we could also do here?
> 
> Another issue is that this won't work for high availability deployments (multiple instances connecting to the same database). There's more places where this will lead to issues, so we can probably ignore this for now.

Probably I'm missing something, but I don't see a major problem on always initializing mediator module, as it will actually initialize the routing key only once and then just query it and store it in the internal variable.

In such case, no new config parameters would be needed. What do you think?

**Comment by TimoGlastra** [comment]
Agree with @genaris here. I think avoiding the config option is worth the little bit of extra overhead.

**Comment by niallshaw-absa** [comment]
Moved to #985

**Pull request closed without merging**